Add fine-tuning secion. Correct multiple inaccuracies. (#915)
tigera/docs: calico-enterprise_versioned_docs/version-3.18/threat/web-application-firewall.mdx

  - name: `openshift`, namespace: `default`
  - name: `gatekeeper-webhook-service`, namespace: `gatekeeper-system`

The rules are not overridden during upgrade, you will have to manage deploying updates of the OWASP CoreRuleSet to your cluster over time. 

If you modify rules, it is recommended to keep your main version in git or similar version control systems. If you enable WAF, modify rules, and then disable WAF again then the rule modifications will be lost. 

:::

## How to

> - [How WAF works](#how-waf-works)

Reviewer: rm

- [Enable WAF on your cluster](#enable-waf)
- [Apply WAF to your services](#apply-waf)
- [View WAF events](#view-waf-events)
- [Fine-tuning your WAF](#fine-tuning-your-waf)
- [Disable WAF feature from your cluster](#disable-waf-feature-from-your-cluster)

> ### How WAF works 

Reviewer: This section should be moved to the Concepts section

kubectl apply -f https://raw.githubusercontent.com/GoogleCloudPlatform/microservices-demo/release/v0.7.0/release/kubernetes-manifests.yaml
```

> #### Enable WAF using CLI

Reviewer: ```suggestion
#### Enable WAF using the CLI
```

##### Enable the Policy Sync API in Felix

> To enable WAF using the CLI, or you have not yet [configured L7 logs](../visibility/elastic/l7/Configure), you must enable the Policy Sync API in Felix. To do this cluster-wide,

Reviewer: ```suggestion
To enable WAF using the CLI, or you have not yet [configured L7 logs](../visibility/elastic/l7/configure.mdx), you must enable the Policy Sync API in Felix. To do this cluster-wide,
```

Reviewer: Re "or you have not yet [configured L7 logs]":

Is this a prerequisite for the entire feature?

Author: no it's not a pre-requisite. If you do have L7 already configured then you just don't need to do this step. I think we can just remove this warning though, there is no harm in trying to run it again if it's already enabled.

    width='600'
  />

### Apply WAF to services

> Now that you have deployed WAF in your cluster, you can select the services you wish to be protected from application layer attacks. 

Reviewer: ```suggestion
Now that you have deployed WAF in your cluster, you can select the services you want to protect from application layer attacks. 
```

If you have deployed the sample application, you can apply WAF on a service associated with your app, as follows:
```bash
kubectl annotate svc frontend -n default --overwrite projectcalico.org/l7-logging=true
```
Alternatively, you can use the Manager UI to apply WAF to the `frontend` service.

In this example, we applied WAF to the `frontend` service. This means that every request that goes through the `frontend` service is inspected.

> However, the traffic is not blocked because the WAF rule is set to `DetectionOnly` by default. You can then begin [fine-tuning your WAF](#fine-tuning-your-waf). 

Reviewer: ```suggestion
However, the traffic is not blocked because the WAF rule is set to `DetectionOnly` by default. You can adjust rules and start blocking traffic by [fine-tuning your WAF](#fine-tuning-your-waf). 
```

You have now applied WAF rule sets to your own services, and note that the traffic that goes through the selected services will be alerted but not blocked by default.

#### Trigger a WAF event

> If you would like to trigger a WAF event for testing purposes, you can simulate an SQL Injection attack inside your cluster by crafting a request with a query string that will trigger WAF. 

Reviewer: ```suggestion
If you would like to trigger a WAF event for testing purposes, you can simulate an SQL injection attack inside your cluster by crafting a request with a query string that will trigger WAF. 
```

> Run this curl command from any pod inside your cluster targeting a service you have selected for WAF protection above e.g. from the demo app above

Reviewer: Would more information about this curl command be helpful? I'm not sure I'd be comfortable pasting this into my teminal.

- [crs-setup.conf](https://github.com/tigera/operator/blob/master/pkg/render/applicationlayer/modsec-core-ruleset/crs-setup.conf)
- [Ruleset files](https://github.com/tigera/operator/tree/master/pkg/render/applicationlayer/modsec-core-ruleset)

There are two ways to edit your rules. 

> 1. Edit the configmap in-place in your cluster using kubectl. The config map combines all the rule files together so you will need to know how to search and find the exact place in the configmap that you want to update. 

Reviewer: ```suggestion
1. Edit the configmap in your cluster using kubectl. The config map combines all the rule files together, so you will need to know how to search and find the exact place in the configmap that you want to update. 
```

> 2. If you want to modify the rules in any meaningful way, it is recommended that you download the files locally and modify before replacing the configmap. In OWASP, it is recommended you disable rules or change configuration by modifying a set of specific metadata files and exclusion files, and not modify any of the main rule files directly. You can, of course, decide to replace the ruleset entirely with your own set of custom rules. This is completely fine once the rules are written in the same SecRule syntax understood by ModSecurity. We recommend you keep a version of your rule files safe in version control like git. Enabling / disabling WAF will override any changes in the ruleset with the default settings. 

Reviewer: ```suggestion
2. If you want to modify the rules in any meaningful way, it is recommended that you download the files locally and modify before replacing the configmap. In OWASP, it is recommended you disable rules or change configuration by modifying a set of specific metadata files and exclusion files, and not modify any of the main rule files directly. You can, of course, decide to replace the ruleset entirely with your own set of custom rules. This is completely fine once the rules are written in the same SecRule syntax understood by ModSecurity. We recommend you keep a version of your rule files safe in version control like git. Enabling or disabling WAF will override any changes in the ruleset with the default settings. 
```

you deploy it to make sure it’s right for your environment.
:::

#### Set WAF rule to block traffic

> By default WAF will not block a request even if it has matching rule violations. The rule engine is set to `DetectionOnly`. You can change this behaviour so that the traffic is blocked instead resulting in a HTTP response code `403 Forbidden` - if the rules triggered score over a certain threshold. 

Reviewer: ```suggestion
By default WAF will not block a request even if it has matching rule violations. The rule engine is set to `DetectionOnly`. You can configure this to block traffic with an HTTP 403 Forbidden response status code when the combined matching rule scores exceed a certain threshold.
```

1. Edit the configmap: 
   ```bash
   kubectl edit cm -n tigera-operator modsecurity-ruleset
   ```
2. Look for `SecRuleEngine DetectionOnly` and change it to `SecRuleEngine On`.

> 3. Save the changes which triggers a rolling update of the L7 pods. 

Reviewer: ```suggestion
3. Save your changes. This triggers a rolling update of the L7 pods. 
```

| Action | Description                                                                                                                                                             | Disruptive? |
| ------ | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ----------- |
| DetectionOnly | Traffic is not denied nor dropped. {{prodname}}  will log events. | No
| On     | Denies HTTP traffic. {{prodname}}  will log the event in Elasticsearch. | Yes          |

> | Off    | Be cautious about using this option. Traffic is not denied nor alerted. |No                                                                      | Yes         |                                         

Reviewer: ```suggestion
| Off    | Be cautious about using this option. Traffic is not denied, and there are no alerts. |No                                                                      | Yes         |                                         
```

#### Change anomaly scoring threshold 

> The default threshold is set to 100, a high value that is unlikely to result in blocked reuqests, even if blocking mode is enabled. This is the recommended starting point from OWASP. You are encouraged to fine-tune and lower this threshold value iteratively until you are happy with the requests being allowed and denied for your cluster. Like most default values you can find this setting in the crs-setup.conf file. You can edit directly in the configmap as follows:

Reviewer: ```suggestion
The default threshold is set to 100, a high value that is unlikely to result in blocked reuqests, even if blocking mode is enabled. This is the recommended starting point from OWASP. You are encouraged to fine-tune and lower this threshold value iteratively until you are happy with the requests being allowed and denied for your cluster. Like most default values you can find this setting in the `crs-setup.conf` file. You can edit directly in the configmap as follows:
```

:::note

> The lower the anomaly scoring threshold value is, the more likely it is that traffic will be denied. 

Reviewer: ```suggestion
The lower the anomaly scoring threshold value, the more likely it is that traffic will be denied. 
```

:::

> 1. Edit the crs-setup.conf file and set the `tx.inbound_anomaly_score_threshold` to the value of your choice: 

Reviewer: ```suggestion
1. Edit the `crs-setup.conf` file and set `tx.inbound_anomaly_score_threshold` to the value of your choice: 
```

   ```bash
     SecAction \
      "id:900110,\
      phase:1,\
      nolog,\
      pass,\
      t:none,\
      setvar:tx.inbound_anomaly_score_threshold=100,\
      setvar:tx.outbound_anomaly_score_threshold=100"
   ```

#### Change other default settings

> The CoreRuleSet has several default values and settings that you can fine-tune. Example: Rule 920420 checks the HTTP request Content-Type header against a list of allowed values. You may want to allow traffic through the WAF that has a certain Content-Type that is not allowed by default, e.g. you may have Content-Types in your micro-services workloads that are not allowed by default. Like most default values you can find the list of allowed Content-Type values in the crs-setup.conf file.

Reviewer: ```suggestion
The CoreRuleSet has several default values and settings that you can fine-tune. Example: Rule 920420 checks the HTTP request Content-Type header against a list of allowed values. You may want to allow traffic through the WAF that has a certain Content-Type that is not allowed by default, e.g. you may have Content-Types in your micro-services workloads that are not allowed by default. Like most default values you can find the list of allowed Content-Type values in the `crs-setup.conf` file.
```

> 1. Edit the crs-setup.conf file and add a new Content-Type by modifying id:900220 and appending to the default list named `tx.allowed_request_content_type`

Reviewer: ```suggestion
1. Edit the `crs-setup.conf` file and add a new Content-Type by modifying `id:900220` and appending to the default list named `tx.allowed_request_content_type`
```

  ```bash
    # Uncomment this rule to change the default.
    #
    SecAction \
     "id:900220,\
      phase:1,\
      nolog,\
      pass,\
      t:none,\
      setvar:'tx.allowed_request_content_type=|application/x-www-form-urlencoded| |multipart/form-data| |multipart/related| |text/xml| |application/xml| |application/soap+xml| |application/json| |application/cloudevents+json| |application/cloudevents-batch+json| |application/grpc|'"
  ```

#### Disable certain rules 

> You may want to disable a rule entirely. For example, if you find the allowed Content-Type rule mentioned above (rule ID 920420) too restrictive you can disable it unconditionally by changing RESPONSE-999-EXCLUSION-RULES-AFTER-CRS.conf.example and renaming that file to remove the .example:

Reviewer: ```suggestion
You may want to disable a rule entirely. For example, if you find the allowed Content-Type rule mentioned above (rule ID 920420) too restrictive, you can disable it unconditionally by changing RESPONSE-999-EXCLUSION-RULES-AFTER-CRS.conf.example and renaming that file to `RESPONSE-999-EXCLUSION-RULES-AFTER-CRS.conf`.
```

1. Edit RESPONSE-999-EXCLUSION-RULES-AFTER-CRS.conf.example and add the rule ID you want to disable

  ```bash
  SecRuleRemoveById 920420
  ```

> 3. Rename RESPONSE-999-EXCLUSION-RULES-AFTER-CRS.conf.example to RESPONSE-999-EXCLUSION-RULES-AFTER-CRS.conf removing the *.example* in order for ModSecurity to load the file

Reviewer: ```suggestion
3. Rename RESPONSE-999-EXCLUSION-RULES-AFTER-CRS.conf.example to RESPONSE-999-EXCLUSION-RULES-AFTER-CRS.conf removing the *.example* in order for ModSecurity to load the file.
```
